File: Darts/src/burlton/dartzee/code/main/DartsMain.java
package burlton.dartzee.code.main;

import javax.swing.UIManager;

import burlton.dartzee.code.screen.DartsApp;
import burlton.dartzee.code.screen.ScreenCache;
import burlton.dartzee.code.utils.DartsDebugExtension;
import burlton.dartzee.code.utils.DartsRegistry;
import burlton.dartzee.code.utils.DesktopDartsClient;
import burlton.dartzee.code.utils.PreferenceUtil;
import burlton.core.code.util.AbstractClient;
import burlton.core.code.util.Debug;
import burlton.core.code.util.DebugUncaughtExceptionHandler;
import burlton.desktopcore.code.util.DialogUtil;
import burlton.core.code.util.OnlineConstants;

public class DartsMain implements DartsRegistry
{
	public static void main(String[] args) 
	{
		Debug.initialise(ScreenCache.getDebugConsole());
		AbstractClient.setInstance(new DesktopDartsClient());
		
		setLookAndFeel();
		
		Debug.setDebugExtension(new DartsDebugExtension());
		Debug.setProductDesc("Darts " + OnlineConstants.DARTS_VERSION_NUMBER);
		Debug.setLogToSystemOut(true);
		
		DartsApp mainScreen = ScreenCache.getMainScreen();
		Thread.setDefaultUncaughtExceptionHandler(new DebugUncaughtExceptionHandler());
		
		AbstractClient.parseProgramArguments(args);
		
		Debug.setSendingEmails(!AbstractClient.devMode);
		
		checkForUpdatesIfRequired();
		
		mainScreen.setVisible(true);
		mainScreen.init();
	}
	
	private static void setLookAndFeel()
	{
		AbstractClient.setOs();
		Debug.append("Initialising Look & Feel - Operating System: " + AbstractClient.operatingSystem);
		if (AbstractClient.isAppleOs())
		{
			setLookAndFeel("javax.swing.plaf.metal");
		}
		else
		{
			setLookAndFeel("javax.swing.plaf.nimbus.NimbusLookAndFeel");
		}
	}
	
	private static void setLookAndFeel(String laf)
	{
		try
		{
			UIManager.setLookAndFeel(laf);
		} 
		catch (Exception e) 
		{
		    Debug.append("Failed to load LookAndFeel " + laf + ". Caught " + e);
		    DialogUtil.showError("Failed to load Look & Feel 'Nimbus'.");
		}
	}
	
	private static void checkForUpdatesIfRequired()
	{
		if (AbstractClient.devMode)
		{
			Debug.append("Not checking for updates as I'm in dev mode");
			return;
		}
		
		if (!PreferenceUtil.getBooleanValue(PREFERENCES_BOOLEAN_CHECK_FOR_UPDATES))
		{
			Debug.append("Not checking for updates as preference is disabled");
			return;
		}
		
		AbstractClient.getInstance().checkForUpdatesIfRequired();
	}
}

File: Darts/src/burlton/dartzee/code/main/DartsMain.kt
package burlton.dartzee.code.main

import burlton.core.code.util.CoreRegistry
import burlton.core.code.util.Debug
import burlton.core.code.util.DebugUncaughtExceptionHandler
import burlton.dartzee.code.`object`.DartsClient
import burlton.dartzee.code.screen.ScreenCache
import burlton.dartzee.code.utils.ClientEmailer
import burlton.dartzee.code.utils.DARTS_VERSION_NUMBER
import burlton.dartzee.code.utils.DartsDebugExtension
import burlton.desktopcore.code.util.DialogUtil
import burlton.desktopcore.code.util.MessageDialogFactory
import javax.swing.JOptionPane
import javax.swing.UIManager

object DartsMain
{
    @JvmStatic
    fun main(args: Array<String>)
    {
        Debug.initialise(ScreenCache.getDebugConsole())
        checkForUserName()
        DialogUtil.init(MessageDialogFactory())

        setLookAndFeel()

        Debug.setDebugExtension(DartsDebugExtension())
        Debug.setProductDesc("Darts $DARTS_VERSION_NUMBER")
        Debug.setLogToSystemOut(true)

        val mainScreen = ScreenCache.getMainScreen()
        Thread.setDefaultUncaughtExceptionHandler(DebugUncaughtExceptionHandler())

        DartsClient.parseProgramArguments(args)

        Debug.setSendingEmails(!DartsClient.devMode)
        ClientEmailer.tryToSendUnsentLogs()

        DartsClient.checkForUpdatesIfRequired()

        mainScreen.isVisible = true
        mainScreen.init()
    }

    private fun checkForUserName()
    {
        var userName: String? = CoreRegistry.instance.get(CoreRegistry.INSTANCE_STRING_USER_NAME, "")
        if (!userName!!.isEmpty())
        {
            return
        }

        Debug.append("Username isn't specified - will prompt for one now")
        while (userName == null || userName.isEmpty())
        {
            userName = JOptionPane.showInputDialog(null, "Please enter your name (for debugging purposes).\nThis will only be asked for once.", "Enter your name")
        }

        CoreRegistry.instance.put(CoreRegistry.INSTANCE_STRING_USER_NAME, userName)

        try
        {
            ClientEmailer.sendClientEmail("Username notification", "$userName has set their username.")
        }
        catch (t: Throwable)
        {
            //If there's no internet connection or Google does something dumb, just log a line
            Debug.append("Caught $t trying to send username notification.")
        }
    }

    private fun setLookAndFeel()
    {
        Debug.append("Initialising Look & Feel - Operating System: ${DartsClient.operatingSystem}")
        if (DartsClient.isAppleOs())
        {
            setLookAndFeel("javax.swing.plaf.metal")
        }
        else
        {
            setLookAndFeel("javax.swing.plaf.nimbus.NimbusLookAndFeel")
        }
    }

    private fun setLookAndFeel(laf: String)
    {
        try
        {
            UIManager.setLookAndFeel(laf)
        }
        catch (e: Exception)
        {
            Debug.append("Failed to load LookAndFeel $laf. Caught $e")
            DialogUtil.showError("Failed to load Look & Feel 'Nimbus'.")
        }

    }
}
